### PR TITLE
[CDAP-15690] Metadata Date Search Implementation

### PR DESCRIPTION
**Pull request by vjusiega**
https://builds.cask.co/browse/CDAP-DRC9426-1/

### CODE REVIEW

File: cdap-standalone/src/main/resources/cdap-site.xml [hunk 1]
```diff
@@ -266,4 +266,9 @@
     </description>
   </property>
 
+  <property>
+    <name>metadata.storage.implementation</name>
```
Comment (CuriousVini) [inline cdap-standalone/src/main/resources/cdap-site.xml]: why is this change needed ?

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java [hunk 2]
```diff
@@ -109,23 +113,78 @@ public static String extractTermValue(String term) {
...
+    return new QueryTerm(term, Qualifier.REQUIRED, QueryTerm.SearchType.DATE, getComparison(valueTerm), date);
+  }
+
+  /**
+   * Parses a string into Unix timestamp if the string format is supported.
+   *
+   * @param term potential date string, with no additional syntax (fields, separators, or comparison operators)
+   * @return Unix timestamp in local time as a Long, or null if the term cannot be parsed
+   */
+  public static Long parseDate(String term) {
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java:171]: this should be marked as `@Nullable` if it can return null.

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java [hunk 3]
```diff
@@ -78,6 +80,28 @@ public QueryTerm(String term, Qualifier qualifier, SearchType searchType, Compar
...
+   */
+  public QueryTerm(String term, Qualifier qualifier, SearchType searchType, Comparison comparison, Long number) {
+    this.term = term;
+    this.qualifier = qualifier;
+    this.searchType = searchType;
+    this.comparison = comparison;
+    if (this.searchType.equals(SearchType.DATE)) {
+      this.date = number;
+    } else {
+      this.date = null;
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java]: you can always initialize date with null and overwrite the value if needed.

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java [hunk 4]
```diff
@@ -1084,10 +1089,66 @@ private QueryBuilder createTermQuery(QueryTerm queryTerm, String textField, Sear
...
+    }
+    if (comp.equals(QueryTerm.Comparison.LESS)) {
+      return QueryBuilders.rangeQuery(field).lte(date - 1d);
+    }
+    if (comp.equals(QueryTerm.Comparison.LESS_OR_EQUAL)) {
+      return QueryBuilders.rangeQuery(field).lte(date);
+    }
+
+    return QueryBuilders.rangeQuery(field).gte(date).lt(date + 1d);
+
```
Comment (CuriousVini) [inline cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: nit: extra line

**Review by CuriousVini — COMMENTED** [review]
a few comments

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java [hunk 5]
```diff
@@ -109,23 +113,78 @@ public static String extractTermValue(String term) {
   }
 
   private static QueryTerm parseQueryTerm(String term) {
-    SearchType parsedSearchType = getSearchType(term);
+    Qualifier parsedQualifier = hasRequirementPrefix(term) ? Qualifier.REQUIRED : Qualifier.OPTIONAL;
+    String parsedTerm = hasRequirementPrefix(term)
+      ? term.substring(1) : term;
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: nit: move this to line 117.

File: cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java [hunk 6]
```diff
@@ -503,6 +515,106 @@ public void testNumericQueries() throws IOException {
     cleanMetadataStorage(records);
   }
 
+  @Test
+  public void testDateParser() throws IOException {
```
Comment (CuriousVini) [inline cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java]: this test should be part of `QueryParserTest`

Comment (CuriousVini) [reply cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java]: Also what would throw an IOException in this case?

File: cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java [hunk 7]
```diff
@@ -503,6 +515,106 @@ public void testNumericQueries() throws IOException {
...
+    String dateString = "2019-08-05";
+    Assert.assertEquals(new Long(1564988400000L), QueryParser.parseDate(dateString));
+  }
+
+  @Test
+  public void testParseDateQuery() throws IOException {
```
Comment (CuriousVini) [inline cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java]: this should also be part of `QueryParserTest`. Also remove `IOException` since its not being thrown by the test

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java [hunk 8]
```diff
@@ -1084,10 +1089,66 @@ private QueryBuilder createTermQuery(QueryTerm queryTerm, String textField, Sear
...
+  /**
+   * Creates a bool query for a date search. If there is no field, creation times and user defined dates are queried.
+   * Otherwise, a field refers to a user defined property, which is searched for with a nested query.
+   *
+   * @param queryTerm QueryTerm object containing the term and its search properties
+   * @param field property field to search in. If undefined search all properties and creation times
+   * @param request the search request
+   * @return a structured query for a date search
+   */
+  private BoolQueryBuilder createDateQuery(QueryTerm queryTerm, String field, SearchRequest request) {
```
Comment (CuriousVini) [inline cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: mark field as nullable

File: cdap-standalone/src/main/resources/cdap-site.xml (hunk 1 above)
Comment (vjusiega) [reply cdap-standalone/src/main/resources/cdap-site.xml]: Oh that was for running elastic search for testing on my computer. I can get rid of that for the next commit.

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 2 above)
Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java:171]: ok, good to know! will update that

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java (hunk 4 above)
Comment (vjusiega) [reply cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: changed in next commit

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java (hunk 8 above)
Comment (vjusiega) [reply cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: got it

File: cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java (hunk 6 above)
Comment (vjusiega) [reply cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java]: You're right, IOException was not being used. Moved the test and removed the IOException.

File: cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java (hunk 7 above)
Comment (vjusiega) [reply cdap-elastic/src/test/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorageTest.java]: Ok

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 5 above)
Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: Do you mean move `SearchType parsedSearchType = getSearchType(term);` to right before line 117? 

Right now I call that line after the if statement in line 120 because parsedSearchType is not needed for the return statement within the if-statement, and so would be computed unnecessarily.

But if that's not a big concern and it helps in readability, I can move it

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java (hunk 3 above)
Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java]: hm, when I try to do that I get an error that date is already initialized, and the issue is coming from date being a final field. I can't figure out how to overwrite it, but please let me know if there is a way!

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 5 above)
Comment (CuriousVini) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: I mean `String parsedTerm = hasRequirementPrefix(term) ? term.substring(1) : term;`

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java (hunk 3 above)
Comment (CuriousVini) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java]: May be we can do this? It is less verbose.  `this.date = this.searchType.equals(SearchType.DATE) ? number : null;`

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 5 above)
Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: Ah I see, ok I can switch that

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java (hunk 3 above)
Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java]: yes that works, thanks!

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java [hunk 9]
```diff
@@ -109,23 +113,78 @@ public static String extractTermValue(String term) {
...
+    if (date == null) {
+      return new QueryTerm(term, qualifier);
+    }
+    // Remove "DATE:" from the term.
+    term = term.substring(5);
+
+    if (term.contains(KEYVALUE_SEPARATOR)) {
+      String[] split = term.split(KEYVALUE_SEPARATOR);
+      if (split.length > 2) {
+        //remove DATE: so that it is not considered as a field name.
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: why are we removing `DATE:` here and in line 158?

Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: line 164 is actually obsolete now because of line 158, so I can remove that if-statement.
I remove 'DATE:' from the query term so that it is not treated as a field when parsed later on. 

Originally I had thought that if a query was of the form `DATE:field:value` then "DATE:" should be removed because a field is defined, but if the form is `DATE:value` then it is possible that the user meant to have "DATE:" as a field and not a keyword. But eventually I decided to not assume that, and instead assume that 'DATE:' always refers to a date search. But let me know your opinions on that!

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java [hunk 10]
```diff
@@ -109,23 +115,73 @@ public static String extractTermValue(String term) {
...
+  /**
+   * Parses a string into Unix timestamp if the string format is supported.
+   *
+   * @param term potential date string, with no additional syntax (fields, separators, or comparison operators)
+   * @return Unix timestamp in local time as a Long, or null if the term cannot be parsed
+   */
+  @Nullable
+  public static Long parseDate(String term) {
+    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+    sdf.setTimeZone(TimeZone.getTimeZone("GTC"));
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: this should be UTC

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java [hunk 11]
```diff
@@ -109,23 +115,73 @@ public static String extractTermValue(String term) {
...
+
+  /**
+   * Parses a string into Unix timestamp if the string format is supported.
+   *
+   * @param term potential date string, with no additional syntax (fields, separators, or comparison operators)
+   * @return Unix timestamp in local time as a Long, or null if the term cannot be parsed
+   */
+  @Nullable
+  public static Long parseDate(String term) {
+    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
```
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: lets call it `format`.

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java [hunk 12]
```diff
@@ -138,4 +141,22 @@ public void testExtractTermValueOfComparisonOperator() {
     Assert.assertEquals("+", QueryParser.extractTermValue(">+"));
     Assert.assertEquals(">", QueryParser.extractTermValue("+key:>"));
   }
+
+  @Test
+  public void testDateParser() {
+    String dateString = "2019-08-05";
+    Assert.assertEquals(new Long(1564963200000L), QueryParser.parseDate(dateString));
```
Comment (CuriousVini) [inline cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: lets use `Long.valueOf`

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java [hunk 13]
```diff
@@ -138,4 +141,22 @@ public void testExtractTermValueOfComparisonOperator() {
...
+  }
+
+  @Test
+  public void testParseDateQuery() {
+    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd");
+    Long date = sdf.parse("2019-08-05", new ParsePosition(0)).getTime();
```
Comment (CuriousVini) [inline cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: nit: unused?

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java [hunk 14]
```diff
@@ -138,4 +141,22 @@ public void testExtractTermValueOfComparisonOperator() {
...
+
+    String query = "DATE:creation_time:<2019-08-05";
+    QueryTerm expected = new QueryTerm("creation_time:<2019-08-05", QueryTerm.Qualifier.REQUIRED,
+                                       QueryTerm.SearchType.DATE, QueryTerm.Comparison.LESS, new Long(1564963200000L));
```
Comment (CuriousVini) [inline cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: `new Long` is not needed.

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 5 above)
Comment (CuriousVini) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: missed?

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java [hunk 15]
```diff
@@ -1084,10 +1089,63 @@ private QueryBuilder createTermQuery(QueryTerm queryTerm, String textField, Sear
...
+   * @param queryTerm    QueryTerm object containing the term and its search properties
+   * @param field        property field to search in
+   * @return a structured range query as a QueryBuilder
+   */
+  private QueryBuilder createDateRangeQuery(QueryTerm queryTerm, String field) {
+    Long date = queryTerm.getDate();
+    QueryTerm.Comparison comp = queryTerm.getComparison();
+
+    if (comp.equals(QueryTerm.Comparison.GREATER)) {
+      return QueryBuilders.rangeQuery(field).gte(date + 1d);
```
Comment (CuriousVini) [inline cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: can we use `.gt` method instead of gte here?

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java [hunk 16]
```diff
@@ -1084,10 +1089,63 @@ private QueryBuilder createTermQuery(QueryTerm queryTerm, String textField, Sear
...
+    }
+    if (comp.equals(QueryTerm.Comparison.GREATER_OR_EQUAL)) {
+      return QueryBuilders.rangeQuery(field).gte(date);
+    }
+    if (comp.equals(QueryTerm.Comparison.LESS)) {
+      return QueryBuilders.rangeQuery(field).lte(date - 1d);
```
Comment (CuriousVini) [inline cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: same question

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java [hunk 17]
```diff
@@ -1084,10 +1089,63 @@ private QueryBuilder createTermQuery(QueryTerm queryTerm, String textField, Sear
...
+   *
+   * @param queryTerm    QueryTerm object containing the term and its search properties
+   * @param field        property field to search in
+   * @return a structured range query as a QueryBuilder
+   */
+  private QueryBuilder createDateRangeQuery(QueryTerm queryTerm, String field) {
+    Long date = queryTerm.getDate();
+    QueryTerm.Comparison comp = queryTerm.getComparison();
+
+    if (comp.equals(QueryTerm.Comparison.GREATER)) {
```
Comment (CuriousVini) [inline cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: lets use switch case similar to `createTermQuery` method

**Review by CuriousVini — COMMENTED** [review]
A few comments

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 10 above)
Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: addressed in next commit

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 11 above)
Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: ok

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java (hunk 13 above)
Comment (vjusiega) [reply cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: you're right. removed now.

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java (hunk 14 above)
Comment (vjusiega) [reply cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: got it, fixed

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java (hunk 12 above)
Comment (vjusiega) [reply cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: got it

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java (hunk 17 above)
Comment (vjusiega) [reply cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: ok

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java (hunk 15 above)
Comment (vjusiega) [reply cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: discussed in chat - gte is necessary to catch dates that are midnight of the next day
added comment about this in code

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java (hunk 16 above)
Comment (vjusiega) [reply cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: added comment to clarify

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java [hunk 18]
```diff
@@ -1084,10 +1089,62 @@ private QueryBuilder createTermQuery(QueryTerm queryTerm, String textField, Sear
...
+   * Otherwise, a field refers to a user defined property, which is searched for with a nested query.
+   *
+   * @param queryTerm QueryTerm object containing the term and its search properties
+   * @param field property field to search in. If undefined search all properties and creation times
+   * @param request the search request
+   * @return a structured query for a date search
+   */
+  private BoolQueryBuilder createDateQuery(QueryTerm queryTerm, @Nullable String field, SearchRequest request) {
+    if (field == null) {
+      QueryBuilder creationDateQuery = createDateRangeQuery(queryTerm, "created").boost(2);
```
Comment (CuriousVini) [inline cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: Lets use static final field `CREATED_FIELD`

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 11 above)
Comment (CuriousVini) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: this means that we will only consider a field as date if its in `yyyy-MM-dd` format right?

Comment (vjusiega) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: yes, that is the current implementation

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java (hunk 18 above)
Comment (vjusiega) [reply cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java]: got it!

**Review by CuriousVini — APPROVED** [review]
LGTM 👍

File: cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java [hunk 19]
```diff
@@ -1129,19 +1129,20 @@ private BoolQueryBuilder createDateQuery(QueryTerm queryTerm, @Nullable String f
...
-        return QueryBuilders.rangeQuery(field).gte(date + 1d);
+        return QueryBuilders.rangeQuery(field).gte(date + oneDay);
       case GREATER_OR_EQUAL:
         return QueryBuilders.rangeQuery(field).gte(date);
       case LESS:
-        return QueryBuilders.rangeQuery(field).lte(date - 1d);
+        return QueryBuilders.rangeQuery(field).lte(date - oneDay);
       case LESS_OR_EQUAL:
-        return QueryBuilders.rangeQuery(field).lte(date);
+        return QueryBuilders.rangeQuery(field).lt(date + oneDay);
```
Comment (CuriousVini) [inline cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java:1143]: why is this change needed?

Comment (vjusiega) [reply cdap-elastic/src/main/java/io/cdap/cdap/metadata/elastic/ElasticsearchMetadataStorage.java:1143]: Once again for the 24h issue, so if creation time is 2019-08-16 at 8am, and we want dates less than or equal to 2019-08-16 then we will still want to return something created on 2019-08-16 at 8am. hope that makes sense!

**Review by CuriousVini — APPROVED** [review]
